docs: update spec for timestamping (#290)
notaryproject/notaryproject: specs/signing-scheme.md

1. Mechanism to establish trust in end users and other entities (like CAs).
1. Set of guarantees available to the verifier apart from integrity and authenticity.

> A signature envelope is restricted to one signing scheme.

Reviewer: > [!IMPORTANT]
> Make it an important block?

1. Mechanism to establish trust in end users and other entities (like CAs).
1. Set of guarantees available to the verifier apart from integrity and authenticity.

> A signature envelope can only specify one Signing Scheme.

Reviewer 2: ```suggestion
A signing envelope is restricted to a one signing scheme.
```

> When Notary Project supports an additional Signing Scheme

Reviewer 2: ```suggestion
When the Notary Project incorporates support for new Signing Scheme
```

>   * Existing signed artifacts MUST be re-signed if they need to be verified using the new signing scheme defined verification process.

Reviewer 2: I think we can omit this because this also apply to existing signing scheme.

>   * Existing clients used by verifying entity MUST be updated to newer versions that support verifying signatures that use the new signing scheme, otherwise the signatures with newer signing schemes which are unknown to existing clients will fail signature verification. Signatures that use older signing schemes which are known to existing clients will continue to verify correctly.

Reviewer 2: ```suggestion
  * For verification of signatures generated using a new signing scheme, it is imperative to update existing signature verification software to newer versions. Failure to do so will result in the verification of these signatures failing. However, signatures generated using older signing schemes will continue to be verified correctly.
```

  * The language of the [Notary Project trust policy](./trust-store-trust-policy.md) in *trustpolicy.json* MAY have breaking changes to support newer concepts/configuration elements introduced by the new signing scheme.

>   The breaking changes are addressed by introducing new major version in the versioned *trustpolicy.json*.

Reviewer 3: I think the breaking changes are not really addressed as the breaking changes are still there compared with older version. Bumping up to a major version is a release strategy that indicates the breaking changes are acceptable.

```suggestion
  The `version` in *trustpolicy.json* will be bumped up to a major version in case any breaking changes introduced.
```

>   * The language of the [Notary Project verification policy](./trust-store-trust-policy.md) in *trustpolicy.json* MAY have breaking changes to support newer concepts/configuration elements introduced by the new signing scheme.

Reviewer 3: In this case, should we consider backward compatible support in Notation?

>   * The language of the [Notary Project verification policy](./trust-store-trust-policy.md) in *trustpolicy.json* MAY have breaking changes to support newer concepts/configuration elements introduced by the new signing scheme.

Reviewer 4: ```suggestion
  * The language of the [Notary Project trust policy](./trust-store-trust-policy.md) in *trustpolicy.json* MAY have breaking changes to support newer concepts/configuration elements introduced by the new signing scheme.
```

>   The breaking changes are addressed by introducing new major version in the versioned *trustpolicy.json*.

Reviewer 2: we should remove this since we dont know if there would be breaking change or not.

Reviewer 2: I see you just rearranged paragraphs

  * The language of the [Notary Project trust policy](./trust-store-trust-policy.md) in *trustpolicy.json* MAY have breaking changes to support newer concepts/configuration elements introduced by the new signing scheme.
  The breaking changes are addressed by introducing new major version in the versioned *trustpolicy.json*.

`notary.x509` - Defines a signing scheme that uses the traditional signing workflow in which an end user generates signatures using X.509 certificates.
## Supported Signing Scheme

`notary.x509.signingAuthority` - Defines a signing scheme in which a `signing authority` generates signatures on behalf of an end user (the signature requestor) using X.509 certificates.
A trusted signing authority is defined as a third party service that is trusted both by the end user (the signature requestor) and verifying entity to generate signatures.
Authorities trusted by users should have mechanisms for validating, logging, monitoring, and auditing access to their systems and publish incidence response plans.

> A certificate authority (CA) will need to demonstrate only validated entities were issued an X.509 certificate that chains to their root, and a time stamp authority (TSA) will need to demonstrate the time-stamp signing keys were only used within their service.

Reviewer 2: Did we remove this?

  * The language of the [Notary Project verification policy](./trust-store-trust-policy.md) in *trustpolicy.json* MAY have breaking changes to support newer concepts/configuration elements introduced by the new signing scheme.
  The breaking changes are addressed by introducing new major version in the versioned *trustpolicy.json*.

## Supported Signing Scheme

Notary Project currently defines the following Signing Schemes:

### `notary.x509`

Defines a signing scheme that uses the traditional signing workflow in which an end user generates signatures using X.509 certificates.

### `notary.x509.signingAuthority`

Defines a signing scheme in which a `signing authority` generates signatures on behalf of an end user (the signature requestor) using X.509 certificates. A trusted signing authority is defined as a third party service that is trusted both by the end user (the signature requestor) and verifying entity to generate signatures.
Authorities trusted by users should have mechanisms for validating, logging, monitoring, and auditing access to their systems and publish incidence response plans.

> A certificate authority (CA) will need to demonstrate only validated entities were issued an X.509 certificate that chains to their root, and a timestamp authority (TSA) will need to demonstrate the timestamp signing keys were only used within their service.

Reviewer 4: This should be a separate paragraph not within `notary.x509.signingAuthority` 
```suggestion
A Certificate Authority (CA) must demonstrate that only validated entities were issued an X.509 certificate that chains to their root. Similarly, a Time-Stamp Authority (TSA) must demonstrate that the time-stamp signing keys were only used within their service. Likewise, a Signing Authority (SA) must demonstrate that signing keys were only used within their service and only validated entities were allowed to generate signatures using the service.
```

Both `notary.x509` and `notary.x509.signingAuthority` signing schemes use the similar signature schema (set of signed and unsigned attributes) with the following differences.

> * `notary.x509` can only use a timestamp countersignature to prove that the signature was generated before the time point been stamped by the `Timestamp Authority (TSA)`.

Reviewer: nit: try not to use the term _only_. _can only use_ is interpreted as _is restricted to use_.
```suggestion
* `notary.x509` can use a timestamp countersignature to prove that the signature was generated before the time point been stamped by the *Timestamp Authority (TSA)*.
```

This is supported through the OPTIONAL *Timestamp Signature* unsigned attribute in the signature envelope. Currently the Notary Project only supports a [RFC3161][ietf-rfc3161] compliant Timestamp for this purpose. When not provided, every certificate in the certificate chain MUST be within its validity period at the time of verification.

> * `notary.x509.signingAuthority` MUST use a date/time attribute that is generated by the signing service as part of the original signature to determine `authentic signing time`.

Reviewer: The term _timestamp_ and _date/time_ are different where _timestamp_ has time zone info but datetime does not.

Both `notary.x509` and `notary.x509.signingAuthority` signing schemes use the similar signature schema (set of signed and unsigned attributes) with the following differences.

> * `notary.x509` can use a timestamp countersignature to prove that the signature was generated before the time point been stamped by the *Timestamp Authority (TSA)*.

Reviewer 2: ```suggestion
* `notary.x509` should use a timestamp countersignature to prove that the signature was generated before the time point been stamped by the *Timestamp Authority (TSA)*.
```

This is supported through the OPTIONAL *Timestamp Signature* unsigned attribute in the signature envelope. Currently the Notary Project only supports a [RFC3161][ietf-rfc3161] compliant Timestamp for this purpose. When not provided, every certificate in the certificate chain MUST be within its validity period at the time of verification.

* `notary.x509.signingAuthority` MUST use a timestamp attribute that is generated by the signing service as part of the original signature to determine `authentic signing time`.

> This is supported through the REQUIRED *Authentic Signing Time* signed attribute in the signature envelope.

Reviewer 2: What happens if signingAuthority doesnt provides `Authentic Signing Time` ?

Both `notary.x509` and `notary.x509.signingAuthority` signing schemes use the similar signature schema (set of signed and unsigned attributes) with the following differences.

> * `notary.x509` does not have the `authentic signing time` concept. It can ONLY use a countersignature (timestamp) to prove that the signature was created on or before the time point been stamped by the `Timestamp Authority (TSA)`.

Reviewer 4: ```suggestion
* `notary.x509` does not have the `authentic signing time` concept. It can ONLY use a countersignature (timestamp) to prove that the signature was created on or before the time point that has been stamped by the `Timestamp Authority (TSA)`.
```

`notary.x509`

* Uses trusts store types Certificate Authority (CA) and Timestamping Authority (TSA) during signature verification.

> The signature is verified against the trust store of type CA, and the *Timestamp Signature* unsigned attribute in the signature envelope is verified against the trust store of type TSA.

Reviewer 4: ```suggestion
The signature is verified against the trust store of type CA, and the *Timestamp Signature* unsigned attribute in the signature envelope is verified against the trust store of type `tsa`.
```
